include/counter.h

```diff
@@ -420,6 +503,38 @@ static inline u32_t z_impl_counter_get_max_relative_alarm(struct device *dev)
 	return api->get_max_relative_alarm(dev);
 }
 
+/**
+ * @brief Return guard period in ticks.
+ *
+ * If late setting detection is used (see @ref Z_COUNTER_ALARM_FLAG_DETECT_LATE)
+ * then certain values are forbidden to be set for alarm timeout because they
+ * will be interpreted as alarm which is in the past and was set too late.
+ *
+ * Late setting detection limits how far in the future alarm can be set. The
+ * limit is calculated as counter top value decreased by the guard period. It
+ * applies to absolute and relative alarms. When late setting detection is
+ * enabled then:
+ * - Relative alarm cannot exceed:
+ * 	top_value - guard_period
+ *
+ * - Absolute value cannot exceed:
+ *	(now + top_value - guard_period) % top_value
+ *
+ *
+ * @param dev		Pointer to the device structure for the driver instance.
+ *
+ * @return Guard period given in ticks.
+ */
+__deprecated __syscall u32_t counter_get_guard_period(struct device *dev);
```

Comment:
How the guard period is being set?

Reply:
it's vendor specific. On nordic i plan to set it in DT, independent per instance.

Reply:
@nordic-krch Where is the example of this?  It was in some PR somewhere, and I remember being very concerned that it should be a KCONFIG setting instead, as it depended on assumptions about the maximum duration that interrupts could be blocked.  I do have a real problem with device-tree settings being application specific (I'm aware of the counter-example about serial console baud rate, but that's less likely to need a per-application override).

Reply:
@pabigot I think we can keep that discussion out of this PR since it will be specific to implementation rather than API. Personally, i don't like DT and like you would prefer to keep everything in kconfig. but it seems that we lean towards using DT for instance configuration (if im wrong it would be good to clarify that). From API usage perspective it does not matter since we access it through api and not through configuration define. Ideally, it should be configured same for all vendors but it's not a must.

Reply:
Huh?  How can we leave it out?  The whole concept of lateness appears to depend on an understanding of guard period.  Last I saw it was defined in elapsed time (ms), which is completely wrong if the counter isn't based on a clock (e.g. I'm counting times a switch has closed).  Maybe it still is, if `tick` here means the same tick used for the system timer, and not a counter advance.

More importantly, you've added API to get the guard period.  It seems natural to expect to be able to set it as well.  I don't think saying "it'll be vendor-specific" is adequate.

Reply:
>I don't think saying "it'll be vendor-specific" is adequate.

That means that it is statically configured and not dynamically. If it is configured during build time then it's not part of API.

Reply:
you may question that it should be dynamically configured and then we can bring it here and analyze in context of that API but then we still need default value that must come from compilation. In that case, i would prefer to keep `set_guard_period` out of API until there is a use case. Current use cases I see don't need setter.

Reply:
> you may question that it should be dynamically configured

Indeed, I was just doing that.

> and then we can bring it here and analyze in context of that API but then we still need default value that must come from compilation. In that case, i would prefer to keep `set_guard_period` out of API until there is a use case. Current use cases I see don't need setter.

I remember https://github.com/zephyrproject-rtos/zephyr/pull/16386#issuecomment-497820500.

Please, let's take the time to fully understand the ramifications and potential use cases of a specific API, rather stopping at the point where it seems to handle one specific case for one vendor.

Reply:
> Please, let's take the time to fully understand the ramifications and potential use cases of a specific API, rather stopping at the point where it seems to handle one specific case for one vendor.

Hm, when one sets the counter it might be set too late. User would probably would like to get notified about this and would like the driver to detect that. Is that one specific use case for one vendor? I think we try to solve this  generic problem here, Do you disagree?